File: Factory method/java/FactoryMethod.java
interface IProduction{
    void release();
}

class Car implements IProduction{
    @Override
    public void release(){
        System.out.println("car");
    }
}

class Truck implements IProduction{
    @Override
    public void release(){
        System.out.println("truck");
    }
}

interface IWorkShop{
    IProduction create();
}

class CarWorkShop implements IWorkShop{
    @Override
    public IProduction create(){
        return new Car();
    }
}

class TruckWorkShop implements IWorkShop{
    @Override
    public IProduction create(){
        return new Truck();
    }
}

public class Main {
    public static void main(String[] args){
        IWorkShop creator = new CarWorkShop();

        IProduction car = creator.create();

        creator = new TruckWorkShop();

        IProduction truck = creator.create();

        car.release();
        truck.release();
    }
}

File: Factory method/c#/FactoryMethod.cs
interface IProduction
{
    void Release();
}

class Car : IProduction
{
    public void Release() => Console.WriteLine("car");
}

class Truck : IProduction
{
    public void Release() => Console.WriteLine("truck");
}

interface IWorkShop
{
    IProduction Create();
}

class CarWorkShop : IWorkShop
{
    public IProduction Create() => new Car();
}

class TruckWorkShop : IWorkShop
{
    public IProduction Create() => new Truck();
}

class Program
{
    static void Main()
    {
        IWorkShop creator = new CarWorkShop();

        IProduction car = creator.Create();

        creator = new TruckWorkShop();

        IProduction truck = creator.Create();
        
        car.Release();
        truck.Release();
    }
}